java-client/session/src/main/java/io/deephaven/client/impl/FilterAdapter.java

```diff
@@ -0,0 +1,149 @@
+//
+// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.client.impl;
+
+import io.deephaven.api.ColumnName;
+import io.deephaven.api.RawString;
+import io.deephaven.api.expression.Expression;
+import io.deephaven.api.expression.Function;
+import io.deephaven.api.expression.Method;
+import io.deephaven.api.filter.*;
+import io.deephaven.proto.backplane.grpc.*;
+
+/**
+ * Converts an Engine API {@link Filter} into a gRPC {@link Condition}.
+ */
+public class FilterAdapter implements Filter.Visitor<Condition> {
```

Comment:
The one caveat that I recently learned about this impl is that it is expected that other kinds of optimizations have already been performed on the queries. There are some cases that it expects to never hit - and therefore is not a round-trip from Filter to Condition (and vice versa). I'm not sure if that matters to you specifically, but if we're increasing the visibility/scope of the adapter we may want to add such a warning to the class javadoc.